esmvalcore/_recipe/recipe.py

```diff
@@ -788,169 +629,94 @@ def update_ancestors(ancestors, step, downstream_settings):
-    """Match a list of input products to output product attributes."""
-    grouped_products = defaultdict(list)
-
-    if not products:
-        return grouped_products
-
-    def get_matching(attributes):
-        """Find the output filename which matches input attributes best."""
-        best_score = 0
-        filenames = []
-        for variable in variables:
-            filename = variable['filename']
-            score = sum(v == variable.get(k) for k, v in attributes.items())
-
-            if score > best_score:
-                best_score = score
-                filenames = [filename]
-            elif score == best_score:
-                filenames.append(filename)
-
-        if not filenames:
-            logger.warning(
-                "Unable to find matching output file for input file %s",
-                filename)
-
-        return filenames
-
-    # Group input files by output file
-    for product in products:
-        matching_filenames = get_matching(product.attributes)
-        for filename in matching_filenames:
-            grouped_products[filename].append(product)
-
-    return grouped_products
-
-
-def _allow_skipping(ancestors, variable, config_user):
+def _allow_skipping(dataset: Dataset):
     """Allow skipping of datasets."""
     allow_skipping = all([
-        config_user.get('skip_nonexistent'),
-        not ancestors,
-        variable['dataset'] != variable.get('reference_dataset'),
+        dataset.session['skip_nonexistent'],
+        dataset.facets['dataset'] != dataset.facets.get('reference_dataset'),
     ])
     return allow_skipping
 
 
-def _get_preprocessor_products(variables, profile, order, ancestor_products,
-                               config_user, name):
+def _set_version(dataset: Dataset, input_datasets: list[Dataset]):
+    """Set the 'version' facet based on derivation input datasets."""
+    versions = set()
+    for in_dataset in input_datasets:
+        in_dataset.set_version()
+        if version := in_dataset.facets.get('version'):
+            if isinstance(version, list):
+                versions.update(version)
+            else:
+                versions.add(version)
+    if versions:
+        version = versions.pop() if len(versions) == 1 else sorted(versions)
+        dataset.set_facet('version', version)
+    for ancillary_ds in dataset.ancillaries:
+        ancillary_ds.set_version()
+
+
+def _get_preprocessor_products(
+    datasets: list[Dataset],
+    profile: dict[str, Any],
+    order: list[str],
+    name: str,
+) -> set[PreprocessorFile]:
     """Get preprocessor product definitions for a set of datasets.
 
     It updates recipe settings as needed by various preprocessors and
     sets the correct ancestry.
     """
     products = set()
-    preproc_dir = config_user['preproc_dir']
 
-    for variable in variables:
-        if variable['frequency'] == 'fx':
-            variable.pop('timerange', None)
-        _update_timerange(variable, config_user)
```

Comment:
The wildcard syntax is really convenient, but I am having some issues when the timerange in given as a wildcard in the recipe you have shared as an example:

```yaml
## same recipe as the example
diagnostics:
  example_diagnostic:
    description: Global mean temperature.
    variables:
      tas:
        mip: Amon
        preprocessor: global_mean
        timerange: '*'
    scripts: null
```
This gives an error when trying to register the ancillary variables (`areacella` in this case) because it tries to guess the timerange in a variable that does not depend on time:

```
Traceback (most recent call last):
  File "/home/b/b381943/ESMValCore/esmvalcore/_main.py", line 481, in run
    fire.Fire(ESMValTool())
  File "/home/b/b381943/mambaforge/envs/core2211/lib/python3.10/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/b/b381943/mambaforge/envs/core2211/lib/python3.10/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/b/b381943/mambaforge/envs/core2211/lib/python3.10/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/home/b/b381943/ESMValCore/esmvalcore/_main.py", line 388, in run
    self._run(recipe, session)
  File "/home/b/b381943/ESMValCore/esmvalcore/_main.py", line 428, in _run
    process_recipe(recipe_file=recipe, session=session)
  File "/home/b/b381943/ESMValCore/esmvalcore/_main.py", line 119, in process_recipe
    recipe = read_recipe_file(recipe_file, session)
  File "/home/b/b381943/ESMValCore/esmvalcore/_recipe/recipe.py", line 72, in read_recipe_file
    return Recipe(raw_recipe, session, recipe_file=filename)
  File "/home/b/b381943/ESMValCore/esmvalcore/_recipe/recipe.py", line 896, in __init__
    self.datasets = Dataset.from_recipe(recipe_file, session)
  File "/home/b/b381943/ESMValCore/esmvalcore/dataset.py", line 128, in from_recipe
    return datasets_from_recipe(recipe, session)  # type: ignore
  File "/home/b/b381943/ESMValCore/esmvalcore/_recipe/to_datasets.py", line 287, in datasets_from_recipe
    variable_datasets = _get_all_datasets_for_variable(
  File "/home/b/b381943/ESMValCore/esmvalcore/_recipe/to_datasets.py", line 248, in _get_all_datasets_for_variable
    for dataset2 in _from_representative_files(dataset1):
  File "/home/b/b381943/ESMValCore/esmvalcore/_recipe/to_datasets.py", line 405, in _from_representative_files
    for repr_ds in repr_dataset.from_files():
  File "/home/b/b381943/ESMValCore/esmvalcore/dataset.py", line 189, in from_files
    ancillaries.extend(ancillary_ds.from_files())
  File "/home/b/b381943/ESMValCore/esmvalcore/dataset.py", line 180, in from_files
    dataset._update_timerange()
  File "/home/b/b381943/ESMValCore/esmvalcore/dataset.py", line 648, in _update_timerange
    intervals = [_get_start_end_date(f.name) for f in dataset.files]
  File "/home/b/b381943/ESMValCore/esmvalcore/dataset.py", line 648, in <listcomp>
    intervals = [_get_start_end_date(f.name) for f in dataset.files]
  File "/home/b/b381943/ESMValCore/esmvalcore/local.py", line 112, in _get_start_end_date
    raise ValueError(f'File {filename} dates do not match a recognized '
ValueError: File areacella_fx_ACCESS-CM2_1pctCO2_r1i1p1f1_gn.nc dates do not match a recognized pattern and time can not be read from the file
```

I think it's because this part of the code that got deleted. I guess that when registering ancillaries there should be a check on the `mip` to avoid trying to guess the timerange given as a wildcard in a variable that does not depend on time.

Reply:
Thanks for testing! I modified the `Dataset._update_timerange` function so it removes the `timerange` facet if the `frequency` is `'fx'` in https://github.com/ESMValGroup/ESMValCore/pull/1609/commits/dec030636353e98abdfa4fdffcf85521d4eb7a7b, so it should work now.